File: src/c/examples/listen.c
/*
 * listen.c — Listen for all incoming frogLink messages.
 *
 * Build: gcc -o listen listen.c ../froglink.c -lpthread
 */

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include "../froglink.h"

static volatile int running = 1;

static void sigint_handler(int sig)
{
    (void)sig;
    running = 0;
}

static void on_message(const char *json_line, void *user_data)
{
    (void)user_data;
    time_t now = time(NULL);
    struct tm *tm = localtime(&now);
    printf("[%02d:%02d:%02d] %s\n", tm->tm_hour, tm->tm_min, tm->tm_sec,
           json_line);
    fflush(stdout);
}

int main(void)
{
    froglink_t fl;
    fl_project_info_t info;

    fl_init(&fl, NULL);

    if (fl_connect(&fl) != FL_OK) {
        printf("Failed to connect. Is the frogLink plugged in?\n");
        return 1;
    }

    fl_error_t err = fl_project(&fl, &info);
    if (err == FL_ERR_NOT_PROVISIONED) {
        printf("frogLink is not provisioned. Run detect first.\n");
        fl_disconnect(&fl);
        return 1;
    }
    if (err != FL_OK) {
        printf("Error querying device.\n");
        fl_disconnect(&fl);
        return 1;
    }

    printf("Connected: %s | %s\n", info.project, info.froglink_name);
    printf("Frogware %s | API %s\n\n", info.sw_version, info.api);

    fl_enable_messages(&fl, true);
    fl_enable_status(&fl, true);

    fl_start_listener(&fl, on_message, NULL);
    printf("Listening for messages (Ctrl+C to stop)...\n\n");

    signal(SIGINT, sigint_handler);

    while (running)
        usleep(100000);

    printf("\nStopping...\n");
    fl_disconnect(&fl);
    return 0;
}

File: src/examples/listen.py
#!/usr/bin/env python3
"""
listen.py — Listen for all incoming frogLink messages.

Connects to the frogLink, enables message and status receiving,
and prints every incoming message until interrupted with Ctrl+C.
Useful for monitoring network activity and debugging.
"""

import json
import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from froglink import FrogLink, FrogLinkError, NotProvisionedError


def on_message(msg):
    """Print each message with a timestamp."""
    ts = time.strftime("%H:%M:%S")
    print(f"[{ts}] {json.dumps(msg)}")


def main():
    link = FrogLink()

    try:
        info = link.connect()
    except NotProvisionedError:
        print("frogLink is not provisioned. Run detect.py for details.")
        sys.exit(1)
    except FrogLinkError as e:
        print(f"Error: {e}")
        sys.exit(1)

    print(f"Connected: {info.get('project', '?')} | {info.get('frogLinkName', '?')}")
    print(f"Frogware {info.get('swVersion', '?')} | API {info.get('api', '?')}")
    print()

    link.enable_messages(True)
    link.enable_status(True)
    link.start_listener(on_message)

    print("Listening for messages (Ctrl+C to stop)...")
    print()

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nStopping...")

    link.disconnect()


if __name__ == "__main__":
    main()
